### PR TITLE
`ModalBarrier`'s gesture recognizer does not call `invokeCallback`

### LINKED ISSUE #125385: `ModalBarrier` has imperfect error message
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

Hi thanks for the wonderful Flutter! It seems that, when `ModalBarrier.onDismiss` callback throws error, Flutter says "Exception was thrown while routing a pointer event" while for normal callbacks it should say "Exception was thrown while handling a gesture". For a new learner of Flutter, they may gets confused what is a "pointer event" while "a gesture" looks more consistent and friendly to new learners.

Indeed, I raise the PR because I hacked the `invokeCallback`, and run into trouble at ModalBarrier because it does not call `invokeCallback` at all. Then I checked the source code and think this is a bit of inconsistency, so I make the PR ;)

Run the test code shown below (at bottom, folded).

### Expected results

After #125386:

```
══╡ EXCEPTION CAUGHT BY GESTURE ╞═══════════════════════════════════════════════════════════════════
The following _Exception was thrown while handling a gesture:
Exception: deliberate

When the exception was thrown, this was the stack:
#0      main.<anonymous closure>.<anonymous closure> (file:///Users/tom/Main/yplusplus/frontend/common_flutter/test/hi.dart:11:28)
#1      ModalBarrier.build.handleDismiss (package:flutter/src/widgets/modal_barrier.dart:120:21)
#2      _AnyTapGestureRecognizer.handleTapUp.<anonymous closure> (package:flutter/src/widgets/modal_barrier.dart:259:49)
#3      GestureRecognizer.invokeCallback (package:flutter/src/gestures/recognizer.dart:255:24)
#4      _AnyTapGestureRecognizer.handleTapUp (package:flutter/src/widgets/modal_barrier.dart:259:5)
#5      BaseTapGestureRecognizer._checkUp (package:flutter/src/gestures/tap.dart:306:5)
#6      BaseTapGestureRecognizer.handlePrimaryPointer (package:flutter/src/gestures/tap.dart:239:7)
#7      PrimaryPointerGestureRecognizer.handleEvent (package:flutter/src/gestures/recognizer.dart:617:9)
#8      PointerRouter._dispatch (package:flutter/src/gestures/pointer_router.dart:98:12)
#9      PointerRouter._dispatchEventToRoutes.<anonymous closure> (package:flutter/src/gestures/pointer_router.dart:143:9)
#10     _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:625:13)
#11     PointerRouter._dispatchEventToRoutes (package:flutter/src/gestures/pointer_router.dart:141:18)
#12     PointerRouter.route (package:flutter/src/gestures/pointer_router.dart:127:7)
#13     GestureBinding.handleEvent (package:flutter/src/gestures/binding.dart:462:19)
#14     GestureBinding.dispatchEvent (package:flutter/src/gestures/binding.dart:442:22)
#15     RendererBinding.dispatchEvent (package:flutter/src/rendering/binding.dart:336:11)
#16     GestureBinding._handlePointerEventImmediately (package:flutter/src/gestures/binding.dart:395:7)
#17     GestureBinding.handlePointerEvent (package:flutter/src/gestures/binding.dart:357:5)
#18     TestWidgetsFlutterBinding.handlePointerEventForSource.<anonymous closure> (package:flutter_test/src/binding.dart:573:42)
#19     TestWidgetsFlutterBinding.withPointerEventSource (package:flutter_test/src/binding.dart:583:11)
#20     TestWidgetsFlutterBinding.handlePointerEventForSource (package:flutter_test/src/binding.dart:573:5)
#21     WidgetTester.sendEventToBinding.<anonymous closure> (package:flutter_test/src/widget_tester.dart:819:15)
#24     TestAsyncUtils.guard (package:flutter_test/src/test_async_utils.dart:71:41)
#25     WidgetTester.sendEventToBinding (package:flutter_test/src/widget_tester.dart:818:27)
#26     TestGesture.up.<anonymous closure> (package:flutter_test/src/test_pointer.dart:554:26)
#29     TestAsyncUtils.guard (package:flutter_test/src/test_async_utils.dart:71:41)
#30     TestGesture.up (package:flutter_test/src/test_pointer.dart:547:27)
#31     WidgetController.tapAt.<anonymous closure> (package:flutter_test/src/controller.dart:511:21)
<asynchronous suspension>
<asynchronous suspension>
(elided 5 frames from dart:async and package:stack_trace)

Handler: "onTapUp"
Recognizer:
  _AnyTapGestureRecognizer#58655
════════════════════════════════════════════════════════════════════════════════════════════════════
-------- the extra --------
extra: {}
-------- Above is an extra dumpErrorToConsole configured from test_config.dart --------
══╡ EXCEPTION CAUGHT BY GESTURE ╞═══════════════════════════════════════════════════════════════════

```

### Actual results

```
══╡ EXCEPTION CAUGHT BY GESTURE LIBRARY ╞═══════════════════════════════════════════════════════════
The following _Exception was thrown while routing a pointer event:
Exception: deliberate

When the exception was thrown, this was the stack:
#0      main.<anonymous closure>.<anonymous closure> (file:///Users/tom/Main/yplusplus/frontend/common_flutter/test/hi.dart:11:28)
#1      ModalBarrier.build.handleDismiss (package:flutter/src/widgets/modal_barrier.dart:120:21)
#2      _AnyTapGestureRecognizer.handleTapUp (package:flutter/src/widgets/modal_barrier.dart:259:17)
#3      BaseTapGestureRecognizer._checkUp (package:flutter/src/gestures/tap.dart:306:5)
#4      BaseTapGestureRecognizer.handlePrimaryPointer (package:flutter/src/gestures/tap.dart:239:7)
#5      PrimaryPointerGestureRecognizer.handleEvent (package:flutter/src/gestures/recognizer.dart:617:9)
#6      PointerRouter._dispatch (package:flutter/src/gestures/pointer_router.dart:98:12)
#7      PointerRouter._dispatchEventToRoutes.<anonymous closure> (package:flutter/src/gestures/pointer_router.dart:143:9)
#8      _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:625:13)
#9      PointerRouter._dispatchEventToRoutes (package:flutter/src/gestures/pointer_router.dart:141:18)
#10     PointerRouter.route (package:flutter/src/gestures/pointer_router.dart:127:7)
#11     GestureBinding.handleEvent (package:flutter/src/gestures/binding.dart:462:19)
#12     GestureBinding.dispatchEvent (package:flutter/src/gestures/binding.dart:442:22)
#13     RendererBinding.dispatchEvent (package:flutter/src/rendering/binding.dart:336:11)
#14     GestureBinding._handlePointerEventImmediately (package:flutter/src/gestures/binding.dart:395:7)
#15     GestureBinding.handlePointerEvent (package:flutter/src/gestures/binding.dart:357:5)
#16     TestWidgetsFlutterBinding.handlePointerEventForSource.<anonymous closure> (package:flutter_test/src/binding.dart:573:42)
#17     TestWidgetsFlutterBinding.withPointerEventSource (package:flutter_test/src/binding.dart:583:11)
#18     TestWidgetsFlutterBinding.handlePointerEventForSource (package:flutter_test/src/binding.dart:573:5)
#19     WidgetTester.sendEventToBinding.<anonymous closure> (package:flutter_test/src/widget_tester.dart:819:15)
#22     TestAsyncUtils.guard (package:flutter_test/src/test_async_utils.dart:71:41)
#23     WidgetTester.sendEventToBinding (package:flutter_test/src/widget_tester.dart:818:27)
#24     TestGesture.up.<anonymous closure> (package:flutter_test/src/test_pointer.dart:554:26)
#27     TestAsyncUtils.guard (package:flutter_test/src/test_async_utils.dart:71:41)
#28     TestGesture.up (package:flutter_test/src/test_pointer.dart:547:27)
#29     WidgetController.tapAt.<anonymous closure> (package:flutter_test/src/controller.dart:511:21)
<asynchronous suspension>
<asynchronous suspension>
(elided 5 frames from dart:async and package:stack_trace)

router: Instance of 'PointerRouter'
route: Closure: (PointerEvent) => void from Function 'handleEvent':.
event: _TransformedPointerUpEvent#75a45(position: Offset(400.0, 300.0))
════════════════════════════════════════════════════════════════════════════════════════════════════
```

### Code sample

<details><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  testWidgets('ModalBarrier', (tester) async {
    final barrierKey = UniqueKey();
    await tester.pumpWidget(MaterialApp(
      home: Scaffold(
        body: ModalBarrier(
          key: barrierKey,
          onDismiss: () => throw Exception('deliberate'),
        ),
      ),
    ));
    await tester.tap(find.byKey(barrierKey));
    await tester.pump();
  });
}
```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[Paste your output here]
```

</details>

### Comparison

For example, say we are using `GestureDetector.onTap` - which should be analogy to `ModalBarrier.onDismiss`:

```
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  testWidgets('ModalBarrier', (tester) async {
    final key = UniqueKey();
    await tester.pumpWidget(MaterialApp(
      home: Scaffold(
        body: GestureDetector(
          key: key,
          onTap: () => throw Exception('deliberate'),
        ),
      ),
    ));
    await tester.tap(find.byKey(key));
    await tester.pump();
  });
}
```

gives

```
══╡ EXCEPTION CAUGHT BY GESTURE ╞═══════════════════════════════════════════════════════════════════
The following _Exception was thrown while handling a gesture:
Exception: deliberate

When the exception was thrown, this was the stack:
...
```

### PR DESCRIPTION
**Pull request by fzyzcjy**
Close #125385 - please see explanations there :)

If you think this PR looks OK, I will polish it (e.g. copy-and-paste-and-modify the tests shown in #125385 into here and make CI pass)

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/widgets/modal_barrier.dart [hunk 1]
```diff
@@ -394,7 +394,7 @@ class _AnyTapGestureRecognizer extends BaseTapGestureRecognizer {
   @protected
   @override
   void handleTapUp({PointerDownEvent? down, PointerUpEvent? up}) {
-    onAnyTapUp?.call();
+    invokeCallback('onTapUp', () => onAnyTapUp?.call());
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/modal_barrier.dart:397]: Looking at the other invocations of invokeCallback, this should probably only be called if onAnyTapUp is non-null?

Comment (goderbauer) [reply packages/flutter/lib/src/widgets/modal_barrier.dart:397]:
```suggestion
    if (onAnyTapUp != null) {
      invokeCallback('onAnyTapUp', onAnyTapUp!);
    }
```

File: packages/flutter/test/widgets/modal_barrier_test.dart [hunk 2]
```diff
@@ -399,6 +399,22 @@ void main() {
...
+            onDismiss: () => throw Exception('deliberate'),
+          ),
+        ),
+      ));
+      await tester.tap(find.byKey(barrierKey));
+      await tester.pump();
+      expect(tester.binding.takeException(), 'TODO_assert_the_error_context');
+    });
+
+
```
Comment (goderbauer) [inline packages/flutter/test/widgets/modal_barrier_test.dart:417]: remove one blank line

File: packages/flutter/test/widgets/modal_barrier_test.dart [hunk 3]
```diff
@@ -399,6 +399,22 @@ void main() {
...
+        home: Scaffold(
+          body: ModalBarrier(
+            key: barrierKey,
+            onDismiss: () => throw Exception('deliberate'),
+          ),
+        ),
+      ));
+      await tester.tap(find.byKey(barrierKey));
+      await tester.pump();
+      expect(tester.binding.takeException(), 'TODO_assert_the_error_context');
```
Comment (goderbauer) [inline packages/flutter/test/widgets/modal_barrier_test.dart:414]: ?

**Comment by fzyzcjy** [comment]
Sure! See my comments above

File: packages/flutter/test/widgets/modal_barrier_test.dart (hunk 2 above)
Comment (fzyzcjy) [reply packages/flutter/test/widgets/modal_barrier_test.dart:417]: (Will do it in batch with my other PRs)

File: packages/flutter/test/widgets/modal_barrier_test.dart (hunk 3 above)
Comment (goderbauer) [reply packages/flutter/test/widgets/modal_barrier_test.dart:414]: Wouldn't the toString of the exception contain something that was added by invokeCallback that you can look for in the test here?

Comment (fzyzcjy) [reply packages/flutter/test/widgets/modal_barrier_test.dart:414]: Hmm,

```
    print(tester.binding.takeException().toString());
```

yields:

```
Exception: deliberate
```

So I guess it does not add anything.

Comment (fzyzcjy) [reply packages/flutter/test/widgets/modal_barrier_test.dart:414]: I propose that, in addition to the existing code:

```dart
  dynamic takeException() {
    assert(inTest);
    final dynamic result = _pendingExceptionDetails?.exception;
    _pendingExceptionDetails = null;
    return result;
  }
  FlutterErrorDetails? _pendingExceptionDetails;
```

We can add:

```dart
FlutterErrorDetails takeErrorDetails() {
    assert(inTest);
    final FlutterErrorDetails result = _pendingExceptionDetails; // <-- NOTE
    _pendingExceptionDetails = null;
    return result;
}
```

Comment (fzyzcjy) [reply packages/flutter/test/widgets/modal_barrier_test.dart:414]: I can create a separate PR for that (given that Flutter guide seems to prefer one PR to do one thing)

Comment (goderbauer) [reply packages/flutter/test/widgets/modal_barrier_test.dart:414]: This shouldn't need new API. Can you get to the error with a custom onError handler? https://master-api.flutter.dev/flutter/foundation/FlutterError/onError.html

Comment (fzyzcjy) [reply packages/flutter/test/widgets/modal_barrier_test.dart:414]: I will try that. Thanks!

**Review by goderbauer — APPROVED** [review]
LGTM

**Comment by goderbauer** [comment]
@fzyzcjy Can you rebase to tip of tree to make the google testing check happy?

**Comment by fzyzcjy** [comment]
Sure!

**Review by chunhtai — APPROVED** [review]
LGTM

**Pull request merged**